substrate/frame/staking-async/rc-client/src/lib.rs

```diff
@@ -637,26 +658,61 @@ pub mod pallet {
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
 			if let Some((report, retries_left)) = OutgoingValidatorSet::<T>::take() {
-				match T::SendToRelayChain::validator_set(report.clone()) {
-					Ok(()) => {
-						// report was sent, all good, it is already deleted.
+				// Determine if we should export the validator set.
+				// Export immediately if:
+				// 1. ValidatorSetExportSession is 0 (immediate export mode)
+				// 2. No era activation has been tracked yet (None = genesis/before first era)
+				let should_export = match LastEraActivationSessionReportEndingIndex::<T>::get() {
+					None => {
+						// No era activation tracked yet, export immediately
+						true
 					},
-					Err(()) => {
-						log!(error, "Failed to send validator set report to relay chain");
-						Self::deposit_event(Event::<T>::Unexpected(
-							UnexpectedKind::ValidatorSetSendFailed,
-						));
-						if let Some(new_retries_left) = retries_left.checked_sub(One::one()) {
-							OutgoingValidatorSet::<T>::put((report, new_retries_left))
+					Some(era_activation_end) => {
+						if T::ValidatorSetExportSession::get() == 0 {
+							// Immediate export mode
+							true
 						} else {
-							Self::deposit_event(Event::<T>::Unexpected(
-								UnexpectedKind::ValidatorSetDropped,
-							));
+							// Check if we've reached the target session offset
+							let last_session_end =
+								LastSessionReportEndingIndex::<T>::get().unwrap_or(0);
+							let current_session = last_session_end.saturating_add(1);
+							let session_offset = current_session.saturating_sub(era_activation_end);
```

Comment:
Minor but you are subtracting a start index (`current_session`) from an end index (`era_activation_end`) here, which is why you have to set `ValidatorSetExportSession`. 

You can also subtract both end indexes, and set `ValidatorSetExportSession` to 4. 

I think this is easier to verify, and makes a good story as: 

- For any era that started at end-index N
- _After receiving_ the end index N+4 we send the next validator set
- Which is received and queued when we receive N+5
- And is activated when we receive N+6

Reply:
Yeah I think we should just use end-index everywhere. Will do!

Reply:
Done here: https://github.com/paritytech/polkadot-sdk/pull/10311/commits/4897e904c89e5c4ecaf35ec270acd88ea09a0c66#diff-a314e246dd6c3592b52f359353e1bf75a084a7a3ae80015a065a6776d14fd05bR678